finality-aleph/src/testing/mocks/validator_network.rs

```diff
@@ -90,3 +112,373 @@ impl<D: Data> MockNetwork<D> {
+        peer_ids.push(id);
+    }
+    // peer main loop
+    // we send random messages to random peers
+    // a message is a number in range 0..n_msg
+    // we also keep a list of messages received at least once
+    // on receiving a message we report the total number of distinct messages received so far
+    // the goal is to receive every message at least once
+    tokio::spawn(async move {
+        let mut received: Vec<bool> = vec![false; n_msg];
+        let mut send_ticker = tokio::time::interval(Duration::from_millis(5));
+        let mut counter: usize = 0;
+        loop {
+            tokio::select! {
+                _ = send_ticker.tick() => {
+                    counter += 1;
+                    // generate random message
+                    let filler_size = if counter % large_message_interval == 0 { 32*1024*1024 } else { 0 };
+                    let decodes = counter % corrupted_message_interval != 0;
+                    let data: MockData = MockData::new(thread_rng().gen_range(0..n_msg) as u32, filler_size, decodes);
+                    // choose a peer
+                    let peer: AuthorityId = peer_ids[thread_rng().gen_range(0..peer_ids.len())].clone();
+                    // send
+                    interface.send(data, peer);
+                },
+                data = interface.next() => {
+                    // receive the message
+                    let data: MockData = data.expect("next should not be closed");
+                    // mark the message as received, we do not care about sender's identity
+                    received[data.data as usize] = true;
+                    // report the number of received messages
+                    report.unbounded_send((our_id.clone(), received.iter().filter(|x| **x).count())).expect("should send");
+                },
+            };
+        }
+    });
+}
+
+pub async fn scenario(
+    n_peers: usize,
+    n_msg: usize,
+    broken_connection_interval: usize,
+    large_message_interval: usize,
+    corrupted_message_interval: usize,
+    status_report_interval: Duration,
+) {
+    // create spawn_handle, we need to keep the task_manager
+    let task_manager =
+        TaskManager::new(Handle::current(), None).expect("should create TaskManager");
+    let spawn_handle = task_manager.spawn_handle();
+    // create peer identities
+    info!(target: "validator-network", "generating keys...");
+    let keys = random_keys(n_peers).await;
+    info!(target: "validator-network", "done");
+    // prepare and run the manager
+    let (mut connection_manager, mut callers, addr) =
+        UnreliableConnectionMaker::new(keys.keys().cloned().collect());
+    tokio::spawn(async move {
+        connection_manager.run(broken_connection_interval).await;
+    });
+    // channel for receiving status updates from spawned peers
+    let (tx_report, mut rx_report) = mpsc::unbounded::<(AuthorityId, usize)>();
+    let mut reports: BTreeMap<AuthorityId, usize> =
```

Comment:
I wonder, why `BTreeMap` here instead of `HashMap`?

Reply:
sorted values in logs

Reply:
Oh, that makes sense. We should use that in status reports in production code then. They are not sorted.

Reply:
not in this PR ofc